lib/internal/http2/core.js

```diff
@@ -1640,7 +1643,17 @@ class Http2Stream extends Duplex {
       }
     }
 
-    this.emit('timeout');
+    // if there is no timeout event listener, destroy session
+    let hasTimeoutCallback = this.emit('timeout');
+    for (const component of this[kStreamEventsComposite]) {
+      if (component.emit('timeout')) {
+        hasTimeoutCallback = true;
+      }
+    }
```

Comment:
I would just check if `if(!this.emit('timeout')) { this.session.destroy() }`. It would need a test on the change in http2 core behavior.

Reply:
I think, this is not what the [docs](https://nodejs.org/api/http2.html#http2_request_settimeout_msecs_callback) are saying. We have 3 EventEmitters (req, res, stream) and if all of them return false, which means "there is no timeout listener" (according to the docs), then we can destroy the session.